### ISSUE TITLE
The `C_GetSlotList()` enumerates slots twice

### ISSUE THREAD

**Issue by Jakuje** [OpenSC/OpenSC#2771]
### Problem Description

The applications usually call `C_GetSlotList()` twice. First is the size inquiry and the second is the actual query for the available slots. This might be both racy and inefficient:
 * the slot listing involves calls to PCSC so we basically call it twice for basically the same information
 * the `C_GetSlotList()` does some allocations that are thrown away when we query only for the size
 * if there is a new reader in between the calls, the calling application might choke on `CKR_BUFFER_TOO_SMALL`

### Proposed Resolution

First obvious proposal would be to cache the `virtual_slots` or `found` lists for these consecutive calls

### Steps to reproduce

Write a program that will wait some time between size inqury and call to `C_GetSlotList()`.
* Remove the reader
* Run the program
* Wait for the first call to `C_GetSlotList()`
* Insert the reader
* Watch the program choke

### Logs
no-logs yet

https://github.com/OpenSC/OpenSC/pull/2740#issuecomment-1534845980

**Comment by frankmorgner** [comment]
I don't think we need to overengineer this. As you've pointed out, the worst thing that could happen is CKR_BUFFER_TOO_SMALL, which applications should be able to cope with. (The *race* condition doesn't lead the program into a blocking condition or similar).

The problem is that we don't know whether or not the two calls to `C_GetSlotList` are dependending on each other or not. Even if the application issues a size inquiry first, the second call may be totally independent from this (especially given the time you need to pass to provoke a race condition). That's why I'd prefer to keep it simple and just do the whole discovery process for every call to `C_GetSlotList`.

**Comment by frankmorgner** [comment]
Having a second look at the implementation, I'd like to note the following observations:
- card_detect_all() is always called in both, query for size and query for list. As consequence, the buffer may be too small if a card is inserted between two calls. virtual_slots are used and updated in both cases. 
- sc_ctx_detect_readers is only called in query for size
- card_detect only reinitalizes the p11card only the card was not present before or the card changed
- The only operation optimisation that I see is the possibility to avoid the `found` buffer allocation in case of a query for size. We will not save any I/O operation.
- 
I don't think that there actually is much to optimise here.

**Comment by Jakuje** [comment]
Thanks for having a look into this. In that case, lets keep it open for some more time if there would be somebody willing to have a look into that or close it if not.

**Comment by frankmorgner** [comment]
Closing this issue due to inactivity. Please re-open the ticket if more input is available.
